File: blobindex/tests/header_fix_test.py
#!/usr/bin/env python3
import struct
import binascii
import blake3

def print_hex(data):
    """Print bytes in hex format"""
    print(binascii.hexlify(data).decode())

# Create MAGIC similar to the one in the codebase
MAGIC = blake3.blake3(b"blobindex").digest()
print("MAGIC: ", end="")
print_hex(MAGIC)

# Define test values (same as in C++ test)
version = 13
volume_prefix = 0x12345678
primary_offset = 4096
secondary_offset = 8192
root_offset = 12288
tail_pos = 16384

print("\n=== CURRENT IMPLEMENTATION (PYTHON STRUCT.PACK) ===")

# Construct the header using struct.pack as in blobindex4.py
header = struct.pack("32sIIQQQQ", 
                    MAGIC,
                    version, 
                    volume_prefix, 
                    primary_offset, 
                    secondary_offset, 
                    root_offset, 
                    tail_pos)

# Print header size
print(f"Header size: {len(header)} bytes")

# Compute checksum
checksum = blake3.blake3(header).digest()

# Print checksum
print("Header checksum: ", end="")
print_hex(checksum)

# Print individual components as packed
print("Individual components as packed:")
print("MAGIC (32 bytes): ", end="")
print_hex(MAGIC)

print("Version (4 bytes): ", end="")
version_bytes = struct.pack("I", version)
print_hex(version_bytes)

print("Volume prefix (4 bytes): ", end="")
prefix_bytes = struct.pack("I", volume_prefix)
print_hex(prefix_bytes)

print("Primary offset (8 bytes): ", end="")
primary_bytes = struct.pack("Q", primary_offset)
print_hex(primary_bytes)

print("Secondary offset (8 bytes): ", end="")
secondary_bytes = struct.pack("Q", secondary_offset)
print_hex(secondary_bytes)

print("Root offset (8 bytes): ", end="")
root_bytes = struct.pack("Q", root_offset)
print_hex(root_bytes)

print("Tail position (8 bytes): ", end="")
tail_bytes = struct.pack("Q", tail_pos)
print_hex(tail_bytes)

print("\n=== FIXED IMPLEMENTATION (COMPATIBLE WITH C++) ===")

# Use a format that matches the C++ implementation
# Note: Python's struct.pack uses native byte order by default
# We need to use '<' to specify little-endian byte order to match C++
header_fixed = struct.pack("<32sIIQQQQ", 
                          MAGIC,
                          version, 
                          volume_prefix, 
                          primary_offset, 
                          secondary_offset, 
                          root_offset, 
                          tail_pos)

# Print header size
print(f"Header size: {len(header_fixed)} bytes")

# Compute checksum
checksum_fixed = blake3.blake3(header_fixed).digest()

# Print checksum
print("Header checksum: ", end="")
print_hex(checksum_fixed)

# Print individual components as packed
print("Individual components as packed:")
print("MAGIC (32 bytes): ", end="")
print_hex(MAGIC)

print("Version (4 bytes): ", end="")
version_bytes = struct.pack("<I", version)
print_hex(version_bytes)

print("Volume prefix (4 bytes): ", end="")
prefix_bytes = struct.pack("<I", volume_prefix)
print_hex(prefix_bytes)

print("Primary offset (8 bytes): ", end="")
primary_bytes = struct.pack("<Q", primary_offset)
print_hex(primary_bytes)

print("Secondary offset (8 bytes): ", end="")
secondary_bytes = struct.pack("<Q", secondary_offset)
print_hex(secondary_bytes)

print("Root offset (8 bytes): ", end="")
root_bytes = struct.pack("<Q", root_offset)
print_hex(root_bytes)

print("Tail position (8 bytes): ", end="")
tail_bytes = struct.pack("<Q", tail_pos)
print_hex(tail_bytes)

# Compare with C++ checksum
print("\n=== COMPARISON ===")
print("Python original checksum: ", end="")
print_hex(checksum)
print("Python fixed checksum: ", end="")
print_hex(checksum_fixed)
print("Expected C++ checksum: 5549bdd001937892edb0d492e049dfa83cb9685f05842c1086c0dc1eb945a16b") 

File: blobindex/tests/header_fix_test.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <string>
#include <array>
#include <cstring>
#include <blake3.h>

// Function to compute BLAKE3 hash
std::array<uint8_t, 32> blake3_hash(const std::vector<uint8_t>& data) {
    std::array<uint8_t, 32> hash;
    blake3_hasher hasher;
    blake3_hasher_init(&hasher);
    blake3_hasher_update(&hasher, data.data(), data.size());
    blake3_hasher_finalize(&hasher, hash.data(), hash.size());
    return hash;
}

// Function to print bytes in hex format
void print_hex(const std::vector<uint8_t>& bytes) {
    for (const auto& byte : bytes) {
        std::cout << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(byte);
    }
    std::cout << std::endl;
}

void print_hex(const std::array<uint8_t, 32>& bytes) {
    for (const auto& byte : bytes) {
        std::cout << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(byte);
    }
    std::cout << std::endl;
}

// Function to print a specific range of bytes in hex format
void print_hex_range(const std::vector<uint8_t>& bytes, size_t start, size_t length) {
    for (size_t i = start; i < start + length && i < bytes.size(); ++i) {
        std::cout << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(bytes[i]);
    }
    std::cout << std::endl;
}

// Struct that matches Python's struct.pack("32sIIQQQQ")
#pragma pack(push, 1)
struct HeaderStruct {
    uint8_t magic[32];
    uint32_t version;
    uint32_t volume_prefix;
    uint64_t primary_offset;
    uint64_t secondary_offset;
    uint64_t root_offset;
    uint64_t tail_pos;
};
#pragma pack(pop)

int main() {
    // Create a MAGIC array similar to the one in the codebase
    std::array<uint8_t, 32> MAGIC;
    blake3_hasher hasher;
    blake3_hasher_init(&hasher);
    blake3_hasher_update(&hasher, "blobindex", 9);
    blake3_hasher_finalize(&hasher, MAGIC.data(), MAGIC.size());
    
    // Print MAGIC
    std::cout << "MAGIC: ";
    print_hex(MAGIC);
    
    std::cout << "\n=== CURRENT IMPLEMENTATION (MANUAL BYTE PACKING) ===" << std::endl;
    
    // Create a header similar to save_index in blobvolume.cpp
    std::vector<uint8_t> header;
    
    // Magic (32 bytes)
    header.insert(header.end(), MAGIC.begin(), MAGIC.end());
    
    // Version (4 bytes) - use consistent byte order
    uint32_t version = 13;
    header.push_back(version & 0xFF);
    header.push_back((version >> 8) & 0xFF);
    header.push_back((version >> 16) & 0xFF);
    header.push_back((version >> 24) & 0xFF);
    
    // Volume prefix (4 bytes) - use consistent byte order
    uint32_t volume_prefix = 0x12345678;
    header.push_back(volume_prefix & 0xFF);
    header.push_back((volume_prefix >> 8) & 0xFF);
    header.push_back((volume_prefix >> 16) & 0xFF);
    header.push_back((volume_prefix >> 24) & 0xFF);
    
    // Primary index offset (8 bytes) - use consistent byte order
    uint64_t primary_offset = 4096;
    header.push_back(primary_offset & 0xFF);
    header.push_back((primary_offset >> 8) & 0xFF);
    header.push_back((primary_offset >> 16) & 0xFF);
    header.push_back((primary_offset >> 24) & 0xFF);
    header.push_back((primary_offset >> 32) & 0xFF);
    header.push_back((primary_offset >> 40) & 0xFF);
    header.push_back((primary_offset >> 48) & 0xFF);
    header.push_back((primary_offset >> 56) & 0xFF);
    
    // Secondary index offset (8 bytes) - use consistent byte order
    uint64_t secondary_offset = 8192;
    header.push_back(secondary_offset & 0xFF);
    header.push_back((secondary_offset >> 8) & 0xFF);
    header.push_back((secondary_offset >> 16) & 0xFF);
    header.push_back((secondary_offset >> 24) & 0xFF);
    header.push_back((secondary_offset >> 32) & 0xFF);
    header.push_back((secondary_offset >> 40) & 0xFF);
    header.push_back((secondary_offset >> 48) & 0xFF);
    header.push_back((secondary_offset >> 56) & 0xFF);
    
    // Root index offset (8 bytes) - use consistent byte order
    uint64_t root_offset = 12288;
    header.push_back(root_offset & 0xFF);
    header.push_back((root_offset >> 8) & 0xFF);
    header.push_back((root_offset >> 16) & 0xFF);
    header.push_back((root_offset >> 24) & 0xFF);
    header.push_back((root_offset >> 32) & 0xFF);
    header.push_back((root_offset >> 40) & 0xFF);
    header.push_back((root_offset >> 48) & 0xFF);
    header.push_back((root_offset >> 56) & 0xFF);
    
    // Tail position (8 bytes) - use consistent byte order
    uint64_t tail_pos = 16384;
    header.push_back(tail_pos & 0xFF);
    header.push_back((tail_pos >> 8) & 0xFF);
    header.push_back((tail_pos >> 16) & 0xFF);
    header.push_back((tail_pos >> 24) & 0xFF);
    header.push_back((tail_pos >> 32) & 0xFF);
    header.push_back((tail_pos >> 40) & 0xFF);
    header.push_back((tail_pos >> 48) & 0xFF);
    header.push_back((tail_pos >> 56) & 0xFF);
    
    // Print header size
    std::cout << "Header size: " << header.size() << " bytes" << std::endl;
    
    // Compute checksum
    std::array<uint8_t, 32> checksum = blake3_hash(header);
    
    // Print checksum
    std::cout << "Header checksum: ";
    print_hex(checksum);
    
    // Print individual components as packed
    std::cout << "Individual components as packed:" << std::endl;
    std::cout << "MAGIC (32 bytes): ";
    print_hex_range(header, 0, 32);
    std::cout << "Version (4 bytes): ";
    print_hex_range(header, 32, 4);
    std::cout << "Volume prefix (4 bytes): ";
    print_hex_range(header, 36, 4);
    std::cout << "Primary offset (8 bytes): ";
    print_hex_range(header, 40, 8);
    std::cout << "Secondary offset (8 bytes): ";
    print_hex_range(header, 48, 8);
    std::cout << "Root offset (8 bytes): ";
    print_hex_range(header, 56, 8);
    std::cout << "Tail position (8 bytes): ";
    print_hex_range(header, 64, 8);
    
    std::cout << "\n=== FIXED IMPLEMENTATION (STRUCT PACKING) ===" << std::endl;
    
    // Create a header using struct packing to match Python's struct.pack
    HeaderStruct header_struct;
    std::memcpy(header_struct.magic, MAGIC.data(), 32);
    header_struct.version = version;
    header_struct.volume_prefix = volume_prefix;
    header_struct.primary_offset = primary_offset;
    header_struct.secondary_offset = secondary_offset;
    header_struct.root_offset = root_offset;
    header_struct.tail_pos = tail_pos;
    
    // Convert struct to vector
    std::vector<uint8_t> header_fixed(sizeof(HeaderStruct));
    std::memcpy(header_fixed.data(), &header_struct, sizeof(HeaderStruct));
    
    // Print header size
    std::cout << "Header size: " << header_fixed.size() << " bytes" << std::endl;
    
    // Compute checksum
    std::array<uint8_t, 32> checksum_fixed = blake3_hash(header_fixed);
    
    // Print checksum
    std::cout << "Header checksum: ";
    print_hex(checksum_fixed);
    
    // Print individual components as packed
    std::cout << "Individual components as packed:" << std::endl;
    std::cout << "MAGIC (32 bytes): ";
    print_hex_range(header_fixed, 0, 32);
    std::cout << "Version (4 bytes): ";
    print_hex_range(header_fixed, 32, 4);
    std::cout << "Volume prefix (4 bytes): ";
    print_hex_range(header_fixed, 36, 4);
    std::cout << "Primary offset (8 bytes): ";
    print_hex_range(header_fixed, 40, 8);
    std::cout << "Secondary offset (8 bytes): ";
    print_hex_range(header_fixed, 48, 8);
    std::cout << "Root offset (8 bytes): ";
    print_hex_range(header_fixed, 56, 8);
    std::cout << "Tail position (8 bytes): ";
    print_hex_range(header_fixed, 64, 8);
    
    return 0;
} 